File: flutter_reach_five_ios/ios/Classes/Converters.swift
import IdentitySdkCore
import IdentitySdkGoogle
import IdentitySdkFacebook

enum CustomFieldError: Error {
    case casting
}

extension CustomField {
    var value: Any {
        switch self {
        case let .string(value): return value
        case let .int(value): return value
        case let .double(value): return value
        case let .bool(value): return value
        case let .array(value): return value
        }
    }
    
    init(value: Any) throws {
        switch value {
        case is String: self = .string(value as! String)
        case is Int: self = .int(value as! Int)
        case is Double: self = .double(value as! Double)
        case is Bool: self = .bool(value as! Bool)
        default: throw CustomFieldError.casting
        }
    }
}

public class Converters {

    static public func parseError(
            reachFiveError: ReachFiveError,
            errorCodesInterface: ErrorCodesInterface,
            defaultFlutterError: FlutterError
    ) -> FlutterError {
        switch reachFiveError {
        case .RequestError(apiError: let apiError):
            if (apiError.errorMessageKey == "error.email.alreadyInUse") {
                return FlutterError(
                        code: errorCodesInterface.emailAlreadyInUseCode,
                        message: apiError.errorUserMsg,
                        details: nil
                )
            }
            if (apiError.errorMessageKey == "error.updateSamePassword") {
                return FlutterError(
                        code: errorCodesInterface.updateSamePassword,
                        message: apiError.errorUserMsg,
                        details: nil
                )
            }
            return defaultFlutterError
        case .AuthFailure(reason: _, apiError: let apiError):
            if (apiError?.errorMessageKey == "error.invalidEmailOrPassword") {
                return FlutterError(
                        code: errorCodesInterface.invalidEmailOrPasswordCode,
                        message: apiError?.errorUserMsg,
                        details: nil
                )
            }
            if (apiError?.errorMessageKey == "error.invalidVerificationCode") {
                return FlutterError(
                        code: errorCodesInterface.invalidVerificationCode,
                        message: apiError?.errorUserMsg,
                        details: nil
                )
            }
            if (apiError?.error == "invalid_grant") {
                return FlutterError(
                        code: errorCodesInterface.invalidGrant,
                        message: apiError?.errorUserMsg,
                        details: nil
                )
            }
            if (apiError?.errorMessageKey == "error.socialAccountEmailAlreadyInUse") {
                return FlutterError(
                        code: errorCodesInterface.socialAccountEmailAlreadyInUse,
                        message: apiError?.errorUserMsg,
                        details: nil
                )
            }
            return defaultFlutterError
        case .AuthCanceled:
            return FlutterError(
                code: errorCodesInterface.userCancelledOrClosedTheWebFlow,
                message: "Auth was cancelled by the user",
                details: nil
            )
        case .TechnicalError(reason: _, apiError: _):
            return defaultFlutterError
        }
    }
    
    static public func authTokenToInterface(
            authToken: AuthToken
        ) -> AuthTokenInterface {
            
            let expriresIn = authToken.expiresIn as NSNumber?
            
            let user = authToken.user != nil
            ? openIdUserToInterface(openIdUser: authToken.user!)
            : nil
            
            return AuthTokenInterface.make(
                withIdToken: authToken.idToken,
                accessToken: authToken.accessToken,
                refreshToken: authToken.refreshToken,
                tokenType: authToken.tokenType,
                expiresIn: expriresIn,
                user: user
            )
        }
    
    static public func authTokenFromInterface(
            authTokenInterface: AuthTokenInterface
        ) -> AuthToken {
            
            let expriresIn = authTokenInterface.expiresIn as? Int
            
            let user = authTokenInterface.user != nil
            ? openIdUserFromInterface(
                openIdUserInterface: authTokenInterface.user!
            )
            : nil
            
            return AuthToken(
                idToken: authTokenInterface.idToken,
                accessToken: authTokenInterface.accessToken,
                refreshToken: authTokenInterface.refreshToken,
                tokenType: authTokenInterface.tokenType,
                expiresIn: expriresIn,
                user: user
            )
        }
    
    static public func openIdUserToInterface(
            openIdUser: OpenIdUser
        ) -> OpenIdUserInterface {
            
            let emailVerified = openIdUser.emailVerified as? NSNumber
            
            let phoneNumberVerified = openIdUser.phoneNumberVerified as? NSNumber
            
            let address = openIdUser.address != nil
            ? addressToInterface(address: openIdUser.address!)
            : nil
            
            return OpenIdUserInterface.make(
                withId: openIdUser.id,
                name: openIdUser.name,
                preferredUsername: openIdUser.preferredUsername,
                givenName: openIdUser.givenName,
                familyName: openIdUser.familyName,
                middleName: openIdUser.middleName,
                nickname: openIdUser.nickname,
                picture: openIdUser.picture,
                website: openIdUser.website,
                email: openIdUser.email,
                emailVerified: emailVerified,
                gender: openIdUser.gender,
                zoneinfo: openIdUser.zoneinfo,
                locale: openIdUser.locale,
                phoneNumber: openIdUser.phoneNumber,
                phoneNumberVerified: phoneNumberVerified,
                address: address,
                birthdate: nil,
                externalId: nil
            )
        }
    
    static public func openIdUserFromInterface(
        openIdUserInterface: OpenIdUserInterface
        ) -> OpenIdUser {

            let address = openIdUserInterface.address != nil
            ? addressFromInterface(
                addressInterface: openIdUserInterface.address!
            )
            : nil
            
            return OpenIdUser(
                id: openIdUserInterface.id,
                name: openIdUserInterface.name,
                preferredUsername: openIdUserInterface.preferredUsername,
                givenName: openIdUserInterface.givenName,
                familyName: openIdUserInterface.familyName,
                middleName: openIdUserInterface.middleName,
                nickname: openIdUserInterface.nickname,
                picture: openIdUserInterface.picture,
                website: openIdUserInterface.website,
                email: openIdUserInterface.email,
                emailVerified: openIdUserInterface.emailVerified?.boolValue,
                gender: openIdUserInterface.gender,
                zoneinfo: openIdUserInterface.zoneinfo,
                locale: openIdUserInterface.locale,
                phoneNumber: openIdUserInterface.phoneNumber,
                phoneNumberVerified: openIdUserInterface.phoneNumberVerified?.boolValue,
                address: address
            )
        }

    static public func addressToInterface(
            address: ProfileAddress
    ) -> AddressInterface {

        AddressInterface.make(
                withFormatted: address.raw,
                streetAddress: address.streetAddress,
                locality: address.locality,
                region: address.region,
                postalCode: address.postalCode,
                country: address.country
        )
    }
    
    static public func addressFromInterface(
            addressInterface: AddressInterface
        ) -> ProfileAddress {

        ProfileAddress(
            title: nil,
            isDefault: nil,
            addressType: nil,
            streetAddress: addressInterface.streetAddress,
            addressComplement: nil,
            locality: addressInterface.locality,
            region: addressInterface.region,
            postalCode: addressInterface.postalCode,
            country: addressInterface.country,
            raw: addressInterface.formatted,
            deliveryNote: nil,
            recipient: nil,
            company: nil,
            phoneNumber: nil,
            customFields: nil
        )
        }
    
    static public func signupRequestFromInterface(
            profileSignupRequestInterface: ProfileSignupRequestInterface
        ) -> ProfileSignupRequest {

            let addresses = profileSignupRequestInterface.addresses?
                .map({ addressRequest in
                    profileAddressFromInterface(
                            profileAddressInterface: addressRequest
                    )
                })
            
            var consents = profileSignupRequestInterface.consents != nil
            ? [String: Consent]()
            : nil
            
            profileSignupRequestInterface.consents?.forEach({
                key, consentInterface in
                consents![key] = consentFromInterface(
                    consentInterface: consentInterface
                )
            })
        
            return ProfileSignupRequest(
                password: profileSignupRequestInterface.password,
                email: profileSignupRequestInterface.email,
                phoneNumber: profileSignupRequestInterface.phoneNumber,
                givenName: profileSignupRequestInterface.givenName,
                middleName: profileSignupRequestInterface.middleName,
                familyName: profileSignupRequestInterface.familyName,
                name: profileSignupRequestInterface.name,
                nickname: profileSignupRequestInterface.nickname,
                birthdate: profileSignupRequestInterface.birthdate,
                profileURL: profileSignupRequestInterface.profileURL,
                picture: profileSignupRequestInterface.picture,
                username: profileSignupRequestInterface.username,
                gender: profileSignupRequestInterface.gender,
                addresses: addresses,
                locale: profileSignupRequestInterface.locale,
                bio: profileSignupRequestInterface.bio,
                consents: consents,
                company: profileSignupRequestInterface.company,
                liteOnly: profileSignupRequestInterface.liteOnly?.boolValue
            )
        }

    static public func addressTypeToInterface(
                addressType: String?
        ) -> ProfileAddressTypeInterface {
            switch addressType {
            case "billing":
                return ProfileAddressTypeInterface.billing
            case "delivery":
                return ProfileAddressTypeInterface.delivery
            default:
                return ProfileAddressTypeInterface.delivery
            }
        }
    
    static public func addressTypeFromInterface(
                addressTypeInterface: ProfileAddressTypeInterface
            ) -> String? {
                switch addressTypeInterface {
                case ProfileAddressTypeInterface.billing:
                    return "billing"
                case ProfileAddressTypeInterface.delivery:
                    return "delivery"
                @unknown default:
                    return nil
                }
            }

    static public func profileAddressToInterface(
            profileAddress: ProfileAddress
    ) -> ProfileAddressInterface {

        let isDefault = profileAddress.isDefault as? NSNumber

        let addressType = addressTypeToInterface(
                addressType: profileAddress.addressType
        )

        return ProfileAddressInterface.make(
                withTitle: profileAddress.title,
                isDefault: isDefault,
                addressType: addressType,
                streetAddress: profileAddress.streetAddress,
                locality: profileAddress.locality,
                region: profileAddress.region,
                postalCode: profileAddress.postalCode,
                country: profileAddress.country,
                raw: profileAddress.raw,
                deliveryNote: profileAddress.deliveryNote,
                recipient: profileAddress.recipient,
                company: profileAddress.company,
                phoneNumber: profileAddress.phoneNumber
        )
    }
    
    static public func profileAddressFromInterface(
            profileAddressInterface: ProfileAddressInterface
        ) -> ProfileAddress {

            let addressType = addressTypeFromInterface(
                addressTypeInterface: profileAddressInterface.addressType
            )
        
            return ProfileAddress(
                title: profileAddressInterface.title,
                isDefault: profileAddressInterface.isDefault?.boolValue,
                addressType: addressType,
                streetAddress: profileAddressInterface.streetAddress,
                addressComplement: nil,

                locality: profileAddressInterface.locality,
                region: profileAddressInterface.region,
                postalCode: profileAddressInterface.postalCode,
                country: profileAddressInterface.country,
                raw: profileAddressInterface.raw,
                deliveryNote: profileAddressInterface.deliveryNote,
                recipient: profileAddressInterface.recipient,
                company: profileAddressInterface.company,
                phoneNumber: profileAddressInterface.phoneNumber,
                customFields: nil
            )
        }

    static public func consentToInterface(
            consent: Consent
    ) -> ConsentInterface {

        ConsentInterface.make(
                withGranted: NSNumber.init(booleanLiteral: consent.granted),
                consentType: consent.consentType,
                date: consent.date
        )
    }
    
    static public func consentFromInterface(
        consentInterface: ConsentInterface
        ) -> Consent {

        Consent(
            granted: consentInterface.granted.boolValue,
            consentType: consentInterface.consentType,
            date: consentInterface.date
        )
    }
    
    static public func providerCreatorFromInterface(
                providerCreatorInterface: ProviderCreatorInterface
            ) -> ProviderCreator? {
                switch providerCreatorInterface.type {
                case ProviderCreatorTypeInterface.google:
                    return GoogleProvider()
                case ProviderCreatorTypeInterface.facebook:
                    return FacebookProvider()
                case ProviderCreatorTypeInterface.webview:
                    return nil
                @unknown default:
                    return nil
                }
            }

    static public func loginSummaryToInterface(
            loginSummary: LoginSummary
    ) -> LoginSummaryInterface {
        let firstLogin = loginSummary.firstLogin != nil ? NSNumber.init(floatLiteral: Double(loginSummary.firstLogin!)) : nil
        let lastLogin = loginSummary.lastLogin != nil ? NSNumber.init(floatLiteral: Double(loginSummary.lastLogin!)) : nil
        let total = loginSummary.total != nil ? NSNumber.init(value: loginSummary.total!) : nil

        return LoginSummaryInterface.make(
                withFirstLogin: firstLogin,
                lastLogin: lastLogin,
                total: total,
                origins: loginSummary.origins,
                devices: loginSummary.devices,
                lastProvider: loginSummary.lastProvider
        )
    }

    static public func loginSummaryFromInterface(
            loginSummaryInterface: LoginSummaryInterface
    ) -> LoginSummary {

        LoginSummary(
                firstLogin: loginSummaryInterface.firstLogin?.intValue,
                lastLogin: loginSummaryInterface.lastLogin?.intValue,
                total: loginSummaryInterface.total?.intValue,
                origins: loginSummaryInterface.origins,
                devices: loginSummaryInterface.devices,
                lastProvider: loginSummaryInterface.lastProvider
        )
    }

    static public func emailsToInterface(
            emails: Emails
    ) -> EmailsInterface {

        EmailsInterface.make(
                withVerified: emails.verified,
                unverified: emails.unverified
        )
    }

    static public func emailsFromInterface(
            emailsInterface: EmailsInterface
    ) -> Emails {

        Emails(
                verified: emailsInterface.verified,
                unverified: emailsInterface.unverified
        )
    }

    static public func profileToInterface(
            profile: Profile
    ) -> ProfileInterface {
        let loginSummary = profile.loginSummary != nil ? loginSummaryToInterface(loginSummary: profile.loginSummary!) : nil
        let emailVerified = profile.emailVerified != nil ? NSNumber.init(booleanLiteral: profile.emailVerified!) : nil
        let emails = profile.emails != nil ? emailsToInterface(emails: profile.emails!) : nil
        let phoneNumberVerified = profile.phoneNumberVerified != nil ? NSNumber.init(booleanLiteral: profile.phoneNumberVerified!) : nil
        let liteOnly = profile.liteOnly != nil ? NSNumber.init(booleanLiteral: profile.liteOnly!) : nil

        let addresses = profile.addresses?
                .map({ profileAddress in
                    profileAddressToInterface(profileAddress: profileAddress)
                })
        
        return ProfileInterface.make(
                withUid: profile.uid,
                givenName: profile.givenName,
                middleName: profile.middleName,
                familyName: profile.familyName,
                name: profile.name,
                nickname: profile.nickname,
                birthdate: profile.birthdate,
                profileURL: profile.profileURL,
                picture: profile.picture,
                externalId: profile.externalId,
                authTypes: profile.authTypes,
                loginSummary: loginSummary,
                username: profile.username,
                gender: profile.gender,
                email: profile.email,
                emailVerified: emailVerified,
                emails: emails,
                phoneNumber: profile.phoneNumber,
                phoneNumberVerified: phoneNumberVerified,
                addresses: addresses,
                locale: profile.locale,
                bio: profile.bio,
                customFields: profile.customFields?.mapValues {customField in customField.value},
                consents: profile.consents?.mapValues {consent in consentToInterface(consent: consent)},
                createdAt: nil,
                updatedAt: nil,
                liteOnly: liteOnly,
                company: nil
        )
    }

    static public func profileFromInterface(
            profileInterface: ProfileInterface
    ) -> Profile {
        let loginSummary = profileInterface.loginSummary != nil ? loginSummaryFromInterface(loginSummaryInterface: profileInterface.loginSummary!) : nil
        let emails = profileInterface.emails != nil ? emailsFromInterface(emailsInterface: profileInterface.emails!) : nil

        let addresses = profileInterface.addresses?
                .map({ profileAddressInterface in
                    profileAddressFromInterface(profileAddressInterface: profileAddressInterface)
                })

        return Profile(
                uid: profileInterface.uid,
                givenName: profileInterface.givenName,
                middleName: profileInterface.middleName,
                familyName: profileInterface.familyName,
                name: profileInterface.name,
                nickname: profileInterface.nickname,
                birthdate: profileInterface.birthdate,
                profileURL: profileInterface.profileURL,
                picture: profileInterface.picture,
                externalId: profileInterface.externalId,
                authTypes: profileInterface.authTypes,
                loginSummary: loginSummary,
                username: profileInterface.username,
                gender: profileInterface.gender,
                email: profileInterface.email,
                emailVerified: profileInterface.emailVerified?.boolValue,
                emails: emails,
                phoneNumber: profileInterface.phoneNumber,
                phoneNumberVerified: profileInterface.phoneNumberVerified?.boolValue,
                addresses: addresses,
                locale: profileInterface.locale,
                bio: profileInterface.bio,
                customFields: try? profileInterface.customFields?.mapValues { customFieldInterface in try CustomField(value: customFieldInterface)},
                consents: profileInterface.consents?.mapValues { consent in consentFromInterface(consentInterface: consent) },
                createdAt: profileInterface.createdAt,
                updatedAt: profileInterface.updatedAt,
                company: profileInterface.company,
                liteOnly: profileInterface.liteOnly?.boolValue
        )
    }
}

File: flutter_reach_five_android/android/src/main/kotlin/tech/bam/flutter_reach_five/android/Converters.kt
package tech.bam.flutter_reach_five.android

import co.reachfive.identity.sdk.core.ProviderCreator
import co.reachfive.identity.sdk.core.models.*
import co.reachfive.identity.sdk.core.models.requests.*
import co.reachfive.identity.sdk.facebook.FacebookProvider
import co.reachfive.identity.sdk.google.GoogleProvider
import co.reachfive.identity.sdk.webview.WebViewProvider

class Converters {
    companion object {
        fun parseError(
        reachFiveError: ReachFiveError,
        errorCodesInterface: ErrorCodesInterface,
        defaultFlutterError: FlutterError
        ): FlutterError {
            if (reachFiveError.data?.errorMessageKey == "error.email.alreadyInUse") {
                val errorCode = errorCodesInterface.emailAlreadyInUseCode
                val errorUsrMessage = reachFiveError.data?.errorUserMsg ?: ""
                return FlutterError(
                    code= errorCode,
                    message= "$errorCode - $errorUsrMessage",
                    details = null
                )
            }
            if (reachFiveError.data?.errorMessageKey == "error.invalidEmailOrPassword") {
                val errorCode = errorCodesInterface.invalidEmailOrPasswordCode
                val errorUsrMessage = reachFiveError.data?.errorUserMsg ?: ""
                return FlutterError(
                    code= errorCode,
                    message= "$errorCode - $errorUsrMessage",
                    details = null
                )
            }
            if (reachFiveError.data?.errorMessageKey == "error.invalidVerificationCode") {
                val errorCode = errorCodesInterface.invalidVerificationCode
                val errorUsrMessage = reachFiveError.data?.errorUserMsg ?: ""
                return FlutterError(
                    code= errorCode,
                    message= "$errorCode - $errorUsrMessage",
                    details = null
                )
            }
            if (reachFiveError.data?.error == "invalid_grant") {
                val errorCode = errorCodesInterface.invalidGrant
                val errorUsrMessage = reachFiveError.data?.errorDescription ?: ""
                return FlutterError(
                    code= errorCode,
                    message= "$errorCode - $errorUsrMessage",
                    details = null
                )
            }
            if (reachFiveError.code == 52001) {
                val errorCode = errorCodesInterface.userCancelledOrClosedTheWebFlow
                val errorMessage = reachFiveError.message
                return FlutterError(
                    code= errorCode,
                    message= "$errorCode - $errorMessage",
                    details = null
                )
            }
            if (reachFiveError.data?.errorMessageKey == "error.socialAccountEmailAlreadyInUse") {
                val errorCode = errorCodesInterface.socialAccountEmailAlreadyInUse
                val errorUsrMessage = reachFiveError.data?.errorDescription ?: ""
                return FlutterError(
                    code= errorCode,
                    message= "$errorCode - $errorUsrMessage",
                    details = null
                )
            }
            return defaultFlutterError
        }

        fun authTokenToInterface(
            authToken: AuthToken
        ): AuthTokenInterface {
            val user = authToken.user

            return AuthTokenInterface(
                    idToken = authToken.idToken,
                    accessToken = authToken.accessToken,
                    refreshToken = authToken.refreshToken,
                    tokenType = authToken.tokenType,
                    expiresIn = authToken.expiresIn?.toLong(),
                    user = if (user != null) openIdUserToInterface(user) else null
            )
        }

        fun authTokenFromInterface(
            authTokenInterface: AuthTokenInterface
        ): AuthToken {
            val user = authTokenInterface.user

            return AuthToken(
                idToken = authTokenInterface.idToken,
                accessToken = authTokenInterface.accessToken,
                refreshToken = authTokenInterface.refreshToken,
                tokenType = authTokenInterface.tokenType,
                expiresIn = authTokenInterface.expiresIn?.toInt(),
                user = if(user != null) openIdUserFromInterface(user) else null
            )
        }

        private fun openIdUserToInterface(
            openIdUser: OpenIdUser
        ): OpenIdUserInterface {
            val address = openIdUser.address

            return OpenIdUserInterface(
                    id = openIdUser.id,
                    name = openIdUser.name,
                    preferredUsername = openIdUser.preferredUsername,
                    givenName = openIdUser.givenName,
                    familyName = openIdUser.familyName,
                    middleName = openIdUser.middleName,
                    nickname = openIdUser.nickname,
                    picture = openIdUser.picture,
                    website = openIdUser.website,
                    email = openIdUser.email,
                    emailVerified = openIdUser.emailVerified,
                    gender = openIdUser.gender,
                    zoneinfo = openIdUser.zoneinfo,
                    locale = openIdUser.locale,
                    phoneNumber = openIdUser.phoneNumber,
                    phoneNumberVerified = openIdUser.phoneNumberVerified,
                    address = if (address != null) addressToInterface(address) else null,
                    birthdate = openIdUser.birthdate,
                    externalId = openIdUser.externalId
            )
        }

        private fun openIdUserFromInterface(
            openIdUserInterface: OpenIdUserInterface
        ): OpenIdUser {
            val address = openIdUserInterface.address

            return OpenIdUser(
                id = openIdUserInterface.id,
                name = openIdUserInterface.name,
                preferredUsername = openIdUserInterface.preferredUsername,
                givenName = openIdUserInterface.givenName,
                familyName = openIdUserInterface.familyName,
                middleName = openIdUserInterface.middleName,
                nickname = openIdUserInterface.nickname,
                picture = openIdUserInterface.picture,
                website = openIdUserInterface.website,
                email = openIdUserInterface.email,
                emailVerified = openIdUserInterface.emailVerified,
                gender = openIdUserInterface.gender,
                zoneinfo = openIdUserInterface.zoneinfo,
                locale = openIdUserInterface.locale,
                phoneNumber = openIdUserInterface.phoneNumber,
                phoneNumberVerified = openIdUserInterface.phoneNumberVerified,
                address = if (address != null) addressFromInterface(address) else null,
                birthdate = openIdUserInterface.birthdate,
                externalId = openIdUserInterface.externalId
            )
        }

        private fun addressToInterface(
            address: Address
        ): AddressInterface {
            return AddressInterface(
                    formatted = address.formatted,
                    streetAddress = address.streetAddress,
                    locality = address.locality,
                    region = address.region,
                    postalCode = address.postalCode,
                    country = address.country,
            )
        }

        private fun addressFromInterface(
            addressInterface: AddressInterface
        ): Address {
            return Address(
                formatted = addressInterface.formatted,
                streetAddress = addressInterface.streetAddress,
                locality = addressInterface.locality,
                region = addressInterface.region,
                postalCode = addressInterface.postalCode,
                country = addressInterface.country,
            )
        }

        fun signupRequestFromInterface(
            profileSignupRequestInterface: ProfileSignupRequestInterface
        ): ProfileSignupRequest {
            val addresses = profileSignupRequestInterface.addresses?.mapNotNull { addressRequest ->
                if (addressRequest != null) profileAddressFromInterface(addressRequest) else null
            }

            val consents = profileSignupRequestInterface.consents?.mapValues { consentRequestMapEntry ->
                val consentRequestMapEntryValue = consentRequestMapEntry.value
                if (consentRequestMapEntryValue != null) consentFromInterface(consentRequestMapEntryValue) else null
            }

            @Suppress("UNCHECKED_CAST")
            return ProfileSignupRequest(
                    password = profileSignupRequestInterface.password,
                    email = profileSignupRequestInterface.email,
                    phoneNumber = profileSignupRequestInterface.phoneNumber,
                    givenName = profileSignupRequestInterface.givenName,
                    middleName = profileSignupRequestInterface.middleName,
                    familyName = profileSignupRequestInterface.familyName,
                    name = profileSignupRequestInterface.name,
                    nickname = profileSignupRequestInterface.nickname,
                    birthdate = profileSignupRequestInterface.birthdate,
                    profileURL = profileSignupRequestInterface.profileURL,
                    picture = profileSignupRequestInterface.picture,
                    username = profileSignupRequestInterface.username,
                    gender = profileSignupRequestInterface.gender,
                    company = profileSignupRequestInterface.company,
                    addresses = addresses,
                    locale = profileSignupRequestInterface.locale,
                    bio = profileSignupRequestInterface.bio,
                    customFields = profileSignupRequestInterface.customFields?.filter {
                        mapEntry -> mapEntry.key == null || mapEntry.value == null } as Map<String, Any>,
                    consents = consents?.filter {
                        mapEntry -> mapEntry.key == null || mapEntry.value == null } as Map<String, Consent>,
                    liteOnly = profileSignupRequestInterface.liteOnly,
            )
        }

        private fun addressTypeToInterface(
            addressType: ProfileAddressType?
        ): ProfileAddressTypeInterface? {
            return when (addressType) {
                ProfileAddressType.billing -> ProfileAddressTypeInterface.BILLING
                ProfileAddressType.delivery -> ProfileAddressTypeInterface.DELIVERY
                null -> null
            }
        }

        private fun addressTypeFromInterface(
            addressTypeInterface: ProfileAddressTypeInterface?
        ): ProfileAddressType? {
            return when (addressTypeInterface) {
                ProfileAddressTypeInterface.BILLING -> ProfileAddressType.billing
                ProfileAddressTypeInterface.DELIVERY -> ProfileAddressType.delivery
                null -> null
            }
        }

        private fun profileAddressToInterface(
            profileAddress: ProfileAddress
        ): ProfileAddressInterface {
            val addressType = addressTypeToInterface(profileAddress.addressType)

            return ProfileAddressInterface(
                    title = profileAddress.title,
                    isDefault = profileAddress.isDefault,
                    addressType = addressType,
                    streetAddress = profileAddress.streetAddress,
                    locality = profileAddress.locality,
                    region = profileAddress.region,
                    postalCode = profileAddress.postalCode,
                    country = profileAddress.country,
                    raw = profileAddress.raw,
                    deliveryNote = profileAddress.deliveryNote,
                    recipient = profileAddress.recipient,
                    company = profileAddress.company,
                    phoneNumber = profileAddress.phoneNumber
            )
        }

        private fun profileAddressFromInterface(
            profileAddressInterface: ProfileAddressInterface
        ): ProfileAddress {
            val addressType = addressTypeFromInterface(profileAddressInterface.addressType)

            return ProfileAddress(
                title = profileAddressInterface.title,
                isDefault = profileAddressInterface.isDefault,
                addressType = addressType,
                streetAddress = profileAddressInterface.streetAddress,
                locality = profileAddressInterface.locality,
                region = profileAddressInterface.region,
                postalCode = profileAddressInterface.postalCode,
                country = profileAddressInterface.country,
                raw = profileAddressInterface.raw,
                deliveryNote = profileAddressInterface.deliveryNote,
                recipient = profileAddressInterface.recipient,
                company = profileAddressInterface.company,
                phoneNumber = profileAddressInterface.phoneNumber
            )
        }

        private fun consentToInterface(
            consent: Consent
        ): ConsentInterface {

            return ConsentInterface(
                    granted = consent.granted,
                    consentType = consent.consentType,
                    date = consent.date,
            )
        }

        private fun consentFromInterface(
            consentInterface: ConsentInterface
        ): Consent {

            return Consent(
                granted = consentInterface.granted,
                consentType = consentInterface.consentType,
                date = consentInterface.date,
            )
        }

        fun providerCreatorFromInterface(
            providerCreatorInterface: ProviderCreatorInterface
        ): ProviderCreator {
            return when (providerCreatorInterface.type) {
                ProviderCreatorTypeInterface.GOOGLE -> GoogleProvider()
                ProviderCreatorTypeInterface.FACEBOOK -> FacebookProvider()
                ProviderCreatorTypeInterface.WEBVIEW -> WebViewProvider()
            }
        }

        private fun loginSummaryToInterface(
            loginSummary: LoginSummary
        ): LoginSummaryInterface {
            return LoginSummaryInterface(
                    firstLogin = loginSummary.firstLogin?.toDouble(),
                    lastLogin = loginSummary.lastLogin?.toDouble(),
                    total = loginSummary.total?.toLong(),
                    origins = loginSummary.origins,
                    devices = loginSummary.devices,
                    lastProvider = loginSummary.lastProvider,
            )
        }

        private fun loginSummaryFromInterface(
            loginSummaryInterface: LoginSummaryInterface
        ): LoginSummary {
            return LoginSummary(
                firstLogin = loginSummaryInterface.firstLogin?.toLong(),
                lastLogin = loginSummaryInterface.lastLogin?.toLong(),
                total = loginSummaryInterface.total?.toInt(),
                origins = loginSummaryInterface.origins?.filterNotNull(),
                devices = loginSummaryInterface.devices?.filterNotNull(),
                lastProvider = loginSummaryInterface.lastProvider
            )
        }

        private fun emailsToInterface(
            emails: Emails
        ): EmailsInterface {
            return EmailsInterface(
                    verified = emails.verified,
                    unverified = emails.unverified
            )
        }

        private fun emailsFromInterface(
            emailsInterface: EmailsInterface
        ): Emails {
            return Emails(
                verified = emailsInterface.verified?.filterNotNull(),
                unverified = emailsInterface.unverified?.filterNotNull()
            )
        }

        fun profileToInterface(
            profile: Profile
        ): ProfileInterface {
            val loginSummary = profile.loginSummary

            val emails = profile.emails

            val addresses = profile.addresses?.map { address ->
                profileAddressToInterface(address)
            }

            val consents = profile.consents?.mapValues { consentMapEntry ->
                consentToInterface(consentMapEntry.value)
            }

            return ProfileInterface(
                    uid = profile.uid,
                    givenName = profile.givenName,
                    middleName = profile.middleName,
                    familyName = profile.familyName,
                    name = profile.name,
                    nickname = profile.nickname,
                    birthdate = profile.birthdate,
                    profileURL = profile.profileURL,
                    picture = profile.picture,
                    externalId = profile.externalId,
                    authTypes = profile.authTypes,
                    loginSummary = if (loginSummary != null) loginSummaryToInterface(loginSummary) else null,
                    username = profile.username,
                    gender = profile.gender,
                    email = profile.email,
                    emailVerified = profile.emailVerified,
                    emails = if (emails != null) emailsToInterface(emails) else null,
                    phoneNumber = profile.phoneNumber,
                    phoneNumberVerified = profile.phoneNumberVerified,
                    addresses = addresses,
                    locale = profile.locale,
                    bio = profile.bio,
                    customFields = profile.customFields?.toMap<String?, Any?>(),
                    consents = consents?.toMap<String?, ConsentInterface?>(),
                    createdAt = profile.createdAt,
                    updatedAt = profile.updatedAt,
                    liteOnly = profile.liteOnly,
                    company = profile.company,

            )
        }

        fun profileFromInterface(
            profileInterface: ProfileInterface
        ): Profile {
            val loginSummaryInterface = profileInterface.loginSummary

            val emailsInterface = profileInterface.emails

            val addresses = profileInterface.addresses?.mapNotNull { address ->
                if (address != null) profileAddressFromInterface(address) else null
            }

            val consents = profileInterface.consents?.mapValues { consentInterfaceMapEntry ->
                val consentInterfaceMapEntryValue = consentInterfaceMapEntry.value
                if (consentInterfaceMapEntryValue != null) consentFromInterface(consentInterfaceMapEntryValue) else null
            }

            @Suppress("UNCHECKED_CAST")
            return Profile(
                uid = profileInterface.uid,
                givenName = profileInterface.givenName,
                middleName = profileInterface.middleName,
                familyName = profileInterface.familyName,
                name = profileInterface.name,
                nickname = profileInterface.nickname,
                birthdate = profileInterface.birthdate,
                profileURL = profileInterface.profileURL,
                picture = profileInterface.picture,
                externalId = profileInterface.externalId,
                authTypes = profileInterface.authTypes?.filterNotNull(),
                loginSummary = if (loginSummaryInterface != null) loginSummaryFromInterface(loginSummaryInterface) else null,
                username = profileInterface.username,
                gender = profileInterface.gender,
                email = profileInterface.email,
                emailVerified = profileInterface.emailVerified,
                emails = if (emailsInterface != null) emailsFromInterface(emailsInterface) else null,
                phoneNumber = profileInterface.phoneNumber,
                phoneNumberVerified = profileInterface.phoneNumberVerified,
                addresses = addresses,
                locale = profileInterface.locale,
                bio = profileInterface.bio,
                customFields = profileInterface.customFields?.filter {
                    mapEntry -> mapEntry.key == null || mapEntry.value == null } as Map<String, Any>,
                consents = consents?.filter {
                    mapEntry -> mapEntry.key == null || mapEntry.value == null } as Map<String, Consent>,
                createdAt = profileInterface.createdAt,
                updatedAt = profileInterface.updatedAt,
                liteOnly = profileInterface.liteOnly,
                company = profileInterface.company
            )
        }
    }
}